### DISCUSSION TITLE
Why doesn't router work when rendered in Jest?

### DISCUSSION THREAD

**Discussion by joonk025** [vercel/next.js#35956] [Help] [answered]
I am rendering a component that uses Next.js router. (useRouter)

When rendering in jsdom using Jest/Testing Library via Testing Library's `render` method, the router always returns `null`.

For example, returning this component in Jest

  
```
import { useRouter } from 'next/router';
const DashboardSidebar: FC<DashboardSidebarProps> = (props) => {
      const { onMobileClose, openMobile } = props;
      const router = useRouter();
      const { user } = useAuth();
    
      useEffect(() => {
        if (openMobile && onMobileClose) {
          onMobileClose();
        }
      }, [router.pathname]);
}
```

will give the error message

```
 TypeError: Cannot read property 'pathname' of null
    
          145 |       onMobileClose();
          146 |     }
        > 147 |   }, [router.pathname]);
```

Same goes for `router.isReady`. router is `null`.

I am assuming that the rendering of the component onto jsdom will run the router, since the router is defined within the component.

I just have to make the router run so that components will render content, and I do not want to test its functionality; I am UI testing, so implementation details are unnecessary.

How does Next.js router work under the hood, and why isn't it running, and how can I make the router run?

"next": "^10.0.1",
"jest": "^27.5.1",
"@testing-library/jest-dom": "^5.16.3",

 "testEnvironment": "jsdom"

**Comment by icyJoseph** [comment, marked as answer]
Hi,

The problem is that at runtime Next.js wraps its Router Context Provider around your app. For debugging in production using React Dev Tools, it is good to be able to figure out where this Provider is, usually the 1st Provider from the top. 


To your question, you need to mock the router, for quick stuff I use https://www.npmjs.com/package/next-router-mock. Because, as you say this makes sure the router works, and then I just test my UI. This is akin to trying to test a component that uses `dispatch` from Redux without having a Redux Provider on the tree. It doesn't matter if its `jest`, the browser, or any other runtime, there must be a Provider somewhere, unless you mock the library.

This discussion contains a few more options: https://github.com/vercel/next.js/discussions/23034.
